src/corelibs/U2Core/src/io/IOAdapterTextStream.cpp

```diff
@@ -0,0 +1,204 @@
+ * MA 02110-1301, USA.
+ */
+
+#include "IOAdapterTextStream.h"
+
+#include <QTextCodec>
+
+#include <U2Core/TextUtils.h>
+#include <U2Core/U2OpStatus.h>
+
+namespace U2 {
+
+///////////////////////////////////////////////
+////      IOAdapterDevice
+///////////////////////////////////////////////
+IOAdapterDevice::IOAdapterDevice(IOAdapter *_ioAdapter, QObject *parent)
+    : QIODevice(parent), ioAdapter(_ioAdapter), currentPos(0) {
+    SAFE_POINT(ioAdapter->isOpen(), "IOAdapter must be opened", );
+    bool isReadable = ioAdapter->isIOModeSupported(IOAdapterMode_Read);
+    bool isWritable = ioAdapter->isIOModeSupported(IOAdapterMode_Write);
+    SAFE_POINT(isReadable || isWritable, "IOAdapter must support read, write or both ops", );
+
+    OpenMode deviceMode(QIODevice::Unbuffered);
+    if (isReadable && isWritable) {
+        deviceMode.setFlag(QIODevice::ReadWrite);
+    } else if (isWritable) {
+        deviceMode.setFlag(QIODevice::WriteOnly);
+    } else {
+        deviceMode.setFlag(QIODevice::ReadOnly);
+    }
+    QIODevice::open(deviceMode);
+}
+
+bool IOAdapterDevice::atEnd() const {
+    return ioAdapter->isEof();
+}
+
+qint64 IOAdapterDevice::readData(char *data, qint64 length) {
+    qint64 bytesRead = ioAdapter->readBlock(data, length);
+    currentPos += bytesRead;
+    return bytesRead;
+}
+
+qint64 IOAdapterDevice::writeData(const char *data, qint64 length) {
+    qint64 bytesWritten = ioAdapter->writeBlock(data, length);
+    currentPos += bytesWritten;
+    return bytesWritten;
+}
+
+qint64 IOAdapterDevice::pos() const {
+    return currentPos;
+}
+
+bool IOAdapterDevice::seek(qint64 seekPos) {
+    qint64 delta = seekPos - currentPos;
+    bool isOk = ioAdapter->skip(delta);
+    if (isOk) {
+        currentPos = seekPos;
+    }
+    return isOk;
+}
+
+///////////////////////////////////////////////
+////      IOAdapterReaderAndWriterBase
+///////////////////////////////////////////////
+IOAdapterReaderAndWriterBase::IOAdapterReaderAndWriterBase(IOAdapter *_ioAdapter, QTextCodec *codec)
+    : ioAdapter(_ioAdapter) {
+    ioDevice.reset(new IOAdapterDevice(ioAdapter));
+    stream.setDevice(ioDevice.data());
+
+    // Set the correct codec, otherwise QTextStream fallbacks to the current locale default (usually UTF-8).
+    if (codec != nullptr) {
+        stream.setCodec(codec);
+    }
+}
+
+GUrl IOAdapterReaderAndWriterBase::getURL() const {
+    return ioAdapter->getURL();
+}
+
+IOAdapterFactory *IOAdapterReaderAndWriterBase::getFactory() const {
+    return ioAdapter->getFactory();
+}
+
+///////////////////////////////////////////////
+////      IOAdapterReader
+///////////////////////////////////////////////
+IOAdapterReader::IOAdapterReader(IOAdapter *ioAdapter)
+    : IOAdapterReaderAndWriterBase(ioAdapter), unreadCharsBufferPos(0) {
+}
+
+IOAdapterReader::~IOAdapterReader() {
+    if (unreadCharsBufferPos < unreadCharsBuffer.length() && ioAdapter->isOpen()) {
```

Comment:
And again - destructior is supposed to clean memory, it's not about doing some post-actions. It's better to connect to [destroyed](https://doc.qt.io/qt-5/qobject.html#destroyed)

Reply:
"And again" - it won't work here. There are no qobjects here and even if they were this very explicit way of lifecycle handling is what needed and it is documented.

Reply:
`IOAdapterReader` is `QObject`, actually, and it's enough to realize my offering. I don't get why you disagree, actually. I mean, I don't get the explanation about lifestyle.

Reply:
You are wrong: IOAdapterReader is not a QObject and your suggestion to always use slot/signals for object destruction is not supported. I use it only when I feel that the pattern reads/works better than the basic plain language.

In this case I want to go low-level as much as possible, stay on the low-level for Reader & Writer. Keeping the destructor to manage internal resources of the stream looks like a reasonable option.

In general I agree - yes: destructors should be avoided. But not in 100% of cases and use of QT slots/signals is not always the best solutions for that.

Reply:
Actually, `IOAdapterReader` is the inheritor of `QObject`. But I don't want to argue anymore, so OK

Reply:
I think you mean IOAdapter, but here it is IOAdapterReader that has no QObject in the hierarchy